base/staticint.jl

```diff
@@ -0,0 +1,357 @@
+
+"""
+    StaticInt(N::Int) -> StaticInt{N}()
+
+A statically sized `Int`. Use `StaticInt(N)` instead of `Val(N)` when you want it to behave like a number.
+"""
+struct StaticInt{N} <: Integer
+    StaticInt{N}() where {N} = new{N::Int}()
+    StaticInt(N::Int) = StaticInt{N}()
+    StaticInt(N::StaticInt) = N
+    StaticInt(::Val{N}) where {N} = StaticInt(N)
+    StaticInt(N) = StaticInt(Int(N))
+end
+
+_dynamic_int(@nospecialize(x::StaticInt))::Int = _dynamic_int(typeof(x))
+_dynamic_int(@nospecialize(x::Type{<:StaticInt}))::Int = x.parameters[1]
+
+const Zero = StaticInt{0}
+const One = StaticInt{1}
+const IntType = Union{Int,StaticInt}
+IntType(x::Integer) = Int(x)
+IntType(x::IntType) = x
+
+convert(::Type{T}, @nospecialize(N::StaticInt)) where {T<:Number} = convert(T, Int(N))
+Bool(x::StaticInt{0}) = false
```

Comment:
We should see how many of these methods are really needed. For example the Real method:
```
Bool(x::Real) = x==0 ? false : x==1 ? true : throw(InexactError(:Bool, Bool, x))
```
should still get fully optimized on StaticInts.

Reply:
I can start cleaning some of this up. Right now it's a slightly outdated hodge-podge of implementation and example use cases.